staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/vault.go

```diff
@@ -0,0 +1,218 @@
+	}
+
+	err = checkConfig(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := newClientWrapper(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vaultEnvelopeService{config: &config, client: client}, nil
+}
+
+func checkConfig(config *EnvelopeConfig) error {
+	if len(config.KeyNames) == 0 {
+		return errors.New("vault provider has no valid key names")
+	}
+
+	if config.Address == "" {
+		return errors.New("vault provider has no valid address")
+	}
+
+	return checkAuthConfig(config)
+}
+
+func checkAuthConfig(config *EnvelopeConfig) error {
+	var count uint
+
+	if config.Token != "" {
+		count++
+	}
+
+	if config.ClientCert != "" || config.ClientKey != "" {
+		if config.ClientCert == "" || config.ClientKey == "" {
+			return errors.New("vault provider has invalid TLS authentication information")
+		}
+		count++
+	}
+
+	if config.RoleID != "" || config.SecretID != "" {
+		if config.RoleID == "" {
+			return errors.New("vault provider has invalid approle authentication information")
+		}
+		count++
+	}
+
+	if count == 0 {
+		return errors.New("vault provider has no authentication information")
+	}
+	if count > 1 {
+		return errors.New("vault provider has more than one authentication information")
+	}
+
+	return nil
+}
+
+type vaultEnvelopeService struct {
+	config *EnvelopeConfig
+	client *clientWrapper
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
+}
+
+func (s *vaultEnvelopeService) Decrypt(data string) ([]byte, error) {
+	// Find the mached key
+	var key string
+	for _, name := range s.config.KeyNames {
+		if strings.HasPrefix(data, name+":") {
+			key = name
+			break
+		}
+	}
+	if key == "" {
+		return nil, errors.New("no matching vault key found")
+	}
+
+	// Replace the key name with "vault:" for Vault transit API
+	cipher := strings.Replace(data, key, vaultPrefix, 1)
+
+	//plain, _, err := s.client.decrypt(key, cipher)
+	plain, err := s.withRefreshToken((*clientWrapper).decrypt, key, cipher)
+	if err != nil {
+		return nil, err
+	}
+
+	return base64.StdEncoding.DecodeString(plain)
+}
+
+func (s *vaultEnvelopeService) Encrypt(data []byte) (string, error) {
+	// Use the frist key to encrypt
+	key := s.config.KeyNames[0]
+	plain := base64.StdEncoding.EncodeToString(data)
+
+	cipher, err := s.withRefreshToken((*clientWrapper).encrypt, key, plain)
+	if err != nil {
+		return "", err
+	}
+
+	// The format of cipher from Vault is "vault:v1:....".
+	// "vault:" is unnecessary, replace it with key name.
+	return strings.Replace(cipher, vaultPrefix, key, 1), nil
+}
+
+// The function type for clientWrapper.encrypt and clientWrapper.decrypt.
+type encryptOrDecryptFunc func(*clientWrapper, string, string) (string, error)
+
+func (s *vaultEnvelopeService) withRefreshToken(f encryptOrDecryptFunc, key, data string) (string, error) {
```

Comment:
We've got similar code in the OIDC plugin and have seen some __really__ bad cases when executing HTTP requests when holding exclusive locks.

I'd expect the mutex to block access to the token, not the actual requests. So something like

```go
mu.Lock()
token := c.token
mu.Unlock()

// make request

// On forbidden
mu.Lock()
if token == c.token { // If not equal, the token's already been rotated
    c.token = c.rotateToken()
}
token = c.token
mu.Unlock()

// retry request.
```

Reply:
Here the rotateToken() itself is an http call, and the rotateToken() and retry_request() need to happen in exclusive lock as in worst case "number of use/ or expiry time" of token can be such that other goroutines consume the token and the goroutine that rotated the token gets forbidden again.
Forbidden case is for the worst case scenarios, We expect vault to be setup with token policies and token renewals jobs such that forbidden case does not get executed normally.

Reply:
> Here the rotateToken() itself is an http call

Right, I'm saying perform this while holding the mutex, not an HTTP request.

> and the rotateToken() and retry_request() need to happen in exclusive lock as in worst case "number of use/ or expiry time" of token can be such that other goroutines consume the token and the goroutine that rotated the token gets forbidden again.

The code snipped above I posted ensures that you only rotate a token if the current token is the one that got a forbidden. If the token's already been rotated it doesn't get rotated again, so the situation you're describing can't happen.

On the contrary, if you're holding a mutex for your HTTP requests, you could start bottle necking all of the concurrent request (what if the vault creds' access has been revoked and you start hitting forbiddens on every request?). We've seen blocking requests long enough result in API server panics.

Overall you want to blocking access to the token not the HTTP request. That's what this suggestion is try to emphasis.

Reply:
@ericchiang Every thread that fails with 403 needs to rotate the token separately, One failed thread cannot rely on the token that was rotated by the previous thread. The token could have got expired, exhausted its number of use or had been revoked explicitly. 

The Logic is as follows
1. A number (say n) of encrypt and decrypt HTTP calls to KMS provider will run concurrently at a given time.
2. Due to token time expiration or token number of use expiration or due to token revocation, one request will fail with 403. This thread will wait for other concurrent HTTP calls to finish. Other concurrent calls will also fail with 403.
3. The first failed thread that acquired exclusive lock will make a token refresh HTTP call and repeat the encrypt or decrypt HTTP call that originally failed and release the lock. 
4. other (n-1) threads will do the same thing as in 3.
5. Once all n threads (plus any new thread that got started in between) that failed pass, the concurrent execution will resume again.
a. This logic needs HTTP requests to be blocked
b. The assumption is n is low for reasons stated in code comments.
c. by good use of token 'time to live' and 'number of use policy', 403 situations can be made less frequent. Explicit revocation of token will incurr 403, but it is not expected to be often.
d. To further avoid the 403 situations, we planned to add (in a separate follow-on PR) a token renew feature which, if configured, will proactively renew the same token before its expiration. This Token renew HTTP request can be made concurrently with encrypt and decrypt HTTP requests.